File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int>res;
        int rb = 0;
        int re = m-1;
        int cb = 0;
        int ce = n-1;
        while(rb<=re && cb<=ce){
            for(int i = cb;i<=ce;i++){
               res.push_back(matrix[rb][i]);
            }
            rb++;
            for(int i = rb;i<=re;i++){
                res.push_back(matrix[i][ce]);
            }
            ce--;
            if(rb<=re){
            for(int i =ce;i>=cb;i--){
                res.push_back(matrix[re][i]);
            }
            }
            re--;
            if(cb<=ce){
            for(int i=re;i>=rb;i--){
                res.push_back(matrix[i][cb]);
            }
            }
            cb++;
        }
        return res;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int r = matrix.length;
        int c = matrix[0].length;

        List<Integer> res = new ArrayList<>();
        int rowBegin = 0;
        int rowEnd = r-1;

        int colBegin = 0;
        int colEnd = c-1;

        while(rowBegin <= rowEnd && colBegin <= colEnd){
            for(int i = colBegin;i<=colEnd ;i++){
                res.add(matrix[rowBegin][i]); 
            }
            rowBegin++;


            for(int i=rowBegin;i<=rowEnd;i++){
                res.add(matrix[i][colEnd]);
            }
            colEnd--;

            if(rowBegin <= rowEnd){

                for(int i=colEnd;i>=colBegin;i--){
                    res.add(matrix[rowEnd][i]);
                }
                rowEnd--;
            }
            

            if(colBegin <= colEnd){
                for(int i=rowEnd;i>=rowBegin;i--){
                    res.add(matrix[i][colBegin]);
                }
                colBegin++;
            }
        }
        return res;
    }
}